### PR TITLE
Listen for removed inputs and unregister them

### PR DESCRIPTION
**Pull request by LiteracyFanatic**
This addresses issue #66 by looping through the `MutationRecord`'s `removedNodes` array and removing entries for any validatable elements from the various state objects. This appears to fix the issue with `isValid` and `isFieldValid` for my test case, but there are a couple parts I'm unsure of:

- I had to add a check for `undefined` validators in `getFormValidationTask` which seems to indicate that `this.formInputs[formUID]` still contains entries for removed inputs even though the `splice` call in `untrackFormInput` should have removed them.
- Not sure if it's cleaner to use the `removed` boolean that's currently being passed to `scan` and its children or to have a separate set of functions instead

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1210,10 +1265,10 @@ export class ValidationService {
     /**
      * Scans the provided root element for any validation directives and attaches behavior to them.
      */
-    scan(root: ParentNode) {
+    scan(root: ParentNode, remove: boolean = false) {
```
Comment (dahlbyk) [inline src/index.ts:1268]: > Not sure if it's cleaner to use the `removed` boolean that's currently being passed to `scan` and its children or to have a separate set of functions instead

IMO the `removed` boolean is fine for the private functions, but not ideal for a public method. Suggest reverting the `scan()` change and adding a `remove(root: ParentNode)` that calls the helpers with `remove = true`.

That said, I don't disagree with your intuition that `remove` isn't particularly clean. You'd have to be careful with `this`, but `scanMessages`/`scanInputs` could receive functions to call in their loop(s):

```diff
     scan(root: ParentNode) {
         this.logger.log('Scanning', root);
-        this.scanMessages(root);
-        this.scanInputs(root);
+        this.scanMessages(root, this.pushValidationMessageSpan);
+        this.scanInputs(root, this.addInput);
     }
```

File: src/index.ts [hunk 2]
```diff
@@ -895,6 +919,20 @@ export class ValidationService {
         this.elementEvents[formUID] = cb;
     }
 
+    private untrackFormInput(form: HTMLFormElement, inputUID: string) {
+        let formUID = this.getElementUID(form);
+        if (!this.formInputs[formUID]) {
+            this.formInputs[formUID] = [];
+        }
```
[the remark this answers is not in the document]
Comment (LiteracyFanatic) [reply src/index.ts:926]: Good point.

File: src/index.ts (hunk 1 above)
Comment (LiteracyFanatic) [reply src/index.ts]: OK, I added a separate `remove` function and left the private functions as is for now.

File: src/index.ts [hunk 3]
```diff
@@ -649,7 +673,10 @@ export class ValidationService {
 
         for (let i = 0; i < formInputUIDs.length; i++) {
             let inputUID = formInputUIDs[i];
-            formValidators.push(this.validators[inputUID]);
+            const validator = this.validators[inputUID];
+            if (validator) {
+                formValidators.push(validator);
+            }
```
Comment (dahlbyk) [inline src/index.ts:679]: Agree it's curious that `inputUID` doesn't seem to have been deleted, but guarding against `undefined` also doesn't both me.

**Review by dahlbyk — APPROVED** [review]
LGTM

**Pull request merged**